accumulator/forestdata.go

```diff
@@ -86,7 +118,1054 @@ func (r *ramForestData) close() {
+	}
+}
+
+// needed as cowForest position translations depend
+// on the correct forestRows
+func (cow *cowForest) setRow(row uint8) {
+	cow.manifest.forestRows = row
+}
+
+func (cow *cowForest) close() {
+	// commit the current forest
+	err := cow.commit()
+	if err != nil {
+		fmt.Printf("cowForest close error:\n%s\n"+
+			"Previously saved forest not overwritten", err)
+	}
+
+	// clean up all old stale files
+	err = cow.clean()
+	if err != nil {
+		panic(err)
+	}
+}
+
+// finds the file and loads it to memory. Does nothing it it's already been
+// loaded.
+func (cow *cowForest) findAndLoad(pos uint64) (uint8, uint64) {
+	treeBlockRow, treeBlockOffset, err := getTreeBlockPos(
+		pos, cow.manifest.forestRows)
+	if err != nil {
+		// TODO better to return err
+		panic(err)
+	}
+
+	treeTableOffset := treeBlockOffset / treeBlockPerTable
+
+	// grab the treeTable location. This is just a number for the .ufod file
+	location := cow.manifest.location[treeBlockRow][treeTableOffset]
+
+	// check if it exists in memory
+	_, found := cow.cachedTreeTables[location]
+
+	// Table is not in memory
+	if !found {
+		// Load the treeTable onto memory. This maps the table to the location
+		err = cow.load(location)
+		if err != nil {
+			// TODO better to return err
+			panic(err)
+		}
+	}
+	return treeBlockRow, treeBlockOffset
+}
+
+// Adds a single new table to the given treeBlockRow in memory
+func (cow *cowForest) newTable(treeBlockRow uint8) {
+	// check if there needs to be a flush
+	if len(cow.cachedTreeTables)+1 > cow.meta.maxCachedTreeTables {
+		cow.flush()
+	}
+
+	if len(cow.manifest.location) <= int(treeBlockRow) {
+		cow.manifest.location = append(cow.manifest.location, []uint64{})
+	}
+	cow.manifest.fileNum++
+	cow.manifest.location[treeBlockRow] = append(
+		cow.manifest.location[treeBlockRow], cow.manifest.fileNum)
+
+	newTable := newTreeTable()
+
+	cow.cachedTreeTables[cow.manifest.fileNum] = newTable
+}
+
+// Update the cowForest num given table location. Returns the new location
+func (cow *cowForest) updateTableNum(
+	location, treeBlockOffset uint64, treeBlockRow uint8) uint64 {
+	// grab the table
+	table := cow.cachedTreeTables[location]
+	if table == nil {
+		panic("fetched table is nil")
+	}
+	tableNew := *table
+
+	treeTableOffset := treeBlockOffset / treeBlockPerTable
+
+	// advance fileNum and set as new file
+	cow.manifest.fileNum++
+	cow.manifest.location[treeBlockRow][treeTableOffset] =
+		cow.manifest.fileNum
+
+	// set as table
+	cow.cachedTreeTables[cow.manifest.fileNum] = &tableNew
+
+	// delete old key
+	delete(cow.cachedTreeTables, location)
+
+	t := cow.cachedTreeTables[cow.manifest.fileNum]
+	if t == nil {
+		panic("fetched t is nil")
+	}
+
+	return cow.manifest.fileNum
+}
+
+// Load will load the existing forest from the disk given a fileNumber
+func (cow *cowForest) load(fileNum uint64) error {
+	// check if there needs to be a flush
+	if len(cow.cachedTreeTables)+1 > cow.meta.maxCachedTreeTables {
+		cow.flush()
```

Comment:
is there maybe a better strategy than flushing everything?

Reply:
Perhaps there is a better way but given 2 gigs for testnet3, it only flushed 3 times during the entire sync so I don't think it's the current bottleneck.

I wanted to get an initial version into master first and then work on improvements. So I think it's not bad putting this on the todo list for now.